server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,382 @@
+
+    /**
+     * Flag to indicate if the request has been aborted, at which point we should stop enqueueing more entries and promptly clean up the
+     * ones being sent.
+     */
+    private final AtomicBoolean isRestResponseFinished = new AtomicBoolean();
+
+    private boolean tryAcquireQueueRef() {
+        return isRestResponseFinished.get() == false && queueRefs.tryIncRef();
+    }
+
+    private void enqueueEntry(ZipEntry zipEntry, ChunkedRestResponseBodyPart firstBodyPart, ActionListener<Void> listener) {
+        if (tryAcquireQueueRef()) {
+            try {
+                entryQueue.add(new ChunkedZipEntry(zipEntry, firstBodyPart, () -> listener.onResponse(null)));
+                if (queueLength.getAndIncrement() == 0) {
+                    final var nextEntry = entryQueue.poll();
+                    assert nextEntry != null;
+                    final var continuation = new QueueConsumer(nextEntry.zipEntry(), nextEntry.firstBodyPart());
+                    assert releasable == null;
+                    releasable = nextEntry.releasable();
+                    final var currentContinuationListener = continuationListener;
+                    continuationListener = new SubscribableListener<>();
+                    if (currentContinuationListener == null) {
+                        final var restResponse = RestResponse.chunked(RestStatus.OK, continuation, this::restResponseFinished);
+                        restResponse.addHeader("content-disposition", Strings.format("attachment; filename=\"%s.zip\"", filename));
+                        restChannel.sendResponse(restResponse);
+                    } else {
+                        currentContinuationListener.onResponse(continuation);
+                    }
+                }
+            } finally {
+                queueRefs.decRef();
+            }
+        } else {
+            listener.onFailure(new AlreadyClosedException("response already closed"));
+        }
+    }
+
+    private void restResponseFinished() {
+        assert Transports.assertTransportThread();
+        if (isRestResponseFinished.compareAndSet(false, true)) {
+            queueRefs.decRef();
+        }
+    }
+
+    private void drainQueue() {
+        assert isRestResponseFinished.get();
+        assert queueRefs.hasReferences() == false;
+        final var taskCount = queueLength.get() + 1;
+        final var releasables = new ArrayList<Releasable>(taskCount);
+        try {
+            if (releasable != null) {
+                releasables.add(releasable);
+                releasable = null;
+            }
+            ChunkedZipEntry entry;
+            while ((entry = entryQueue.poll()) != null) {
+                releasables.add(entry.releasable());
+            }
+            assert entryQueue.isEmpty() : entryQueue.size(); // no concurrent adds
+            assert releasables.size() <= taskCount : taskCount + " vs " + releasables.size();
```

Comment:
I think that would be a weaker assertion, since it would also drain any items to be concurrently added to the queue. We really want to check that we didn't drain any more items than were present when `drainQueue()` was called.

Reply:
> it would also drain any items to be concurrently added to the queue

I don't think that's possible since we assert `queueRefs.hasReferences() == false` a few lines above? My thinking is that `<=` is not accurate since `<` can allow many values.